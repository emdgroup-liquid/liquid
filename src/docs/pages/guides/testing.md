---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

