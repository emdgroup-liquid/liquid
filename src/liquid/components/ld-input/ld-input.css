/*
:where(.ld-input) {
  width: 12rem;
}
*/

.ld-input {
  --ld-input-padding-x: 0.625rem;
  --ld-input-padding-top: 0.625rem;
  --ld-input-padding-bottom: 0.6875rem;
  --ld-input-max-height: 2.5rem;
  display: inline-flex;
  align-items: center;
  background-color: var(--ld-col-bg-g);
  color: var(--ld-col-rblck-default);
  box-shadow: inset 0 0 0 0.1rem var(--ld-col-rblck1);
  max-width: 100%;
  border-radius: var(--ld-sp-4);
  line-height: 1;

  > :where(:not(input)) {
    user-select: none;
  }

  > :where(:not(input):first-child) {
    margin-left: var(--ld-sp-12);
  }

  > :where(:not(input):last-child) {
    margin-right: var(--ld-sp-12);
  }

  > input {
    padding: var(--ld-input-padding-top) var(--ld-input-padding-x)
      var(--ld-input-padding-bottom);
    font: var(--ld-typo-body-m);
    line-height: 1;
    background-color: transparent;
    width: 100%;
    border: 0;
    border-radius: var(--ld-sp-4);
    max-height: var(--ld-input-max-height);
    -webkit-appearance: none;

    &::placeholder {
      color: var(--ld-col-rblck4);
    }

    &[type='search'] {
      &::-webkit-search-decoration,
      &::-webkit-search-cancel-button,
      &::-webkit-search-results-button,
      &::-webkit-search-results-decoration {
        -webkit-appearance: none;
      }
    }
  }

  > ld-button,
  > .ld-button {
    background-color: transparent;

    &:first-child,
    &:first-child > button {
      margin-left: 0;

      &:not(.ld-button--sm) {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      }

      &.ld-button--sm {
        margin: var(--ld-sp-4);
        margin-right: 0;
      }
    }

    &:last-child,
    &:last-child > button {
      margin-right: 0;

      &:not(.ld-button--sm) {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }

      &.ld-button--sm {
        margin: var(--ld-sp-4);
        margin-left: 0;
      }
    }

    &[mode='ghost'],
    &.ld-button--ghost {
      &:first-child {
        margin-right: calc(-1 * var(--ld-input-padding-x));
      }

      &:last-child {
        margin-left: calc(-1 * var(--ld-input-padding-x));
      }
    }
  }
  > ld-button.hydrated > button.ld-button--ghost,
  > button.ld-button.ld-button--ghost {
    @media (hover: hover) {
      &:hover {
        background-color: transparent;
      }
    }
    &:active,
    &:active:focus-visible,
    &:focus:focus-visible {
      background-color: transparent;
    }
  }

  > ld-icon,
  > .ld-icon {
    color: var(--ld-col-rb-default);
    display: inline-flex;
  }

  .ld-theme-solvent &,
  .ld-theme-bubblegum &,
  .ld-theme-shake & {
    > ld-icon,
    > .ld-icon {
      color: var(--ld-col-rp-default);
    }
  }

  .ld-theme-tea & {
    > ld-icon,
    > .ld-icon {
      color: var(--ld-col-rg-default);
    }
  }

  .ld-label & {
    width: 100%;
  }
}

.ld-input--light {
  background-color: var(--ld-col-wht);
}

:where(.ld-input:not([disabled]:not([aria-disabled]))) {
  &:hover {
    box-shadow: inset 0 0 0 0.1rem var(--ld-col-rblck2);
  }
}

:where(.ld-input) {
  &[disabled],
  &[aria-disabled] {
    box-shadow: inset 0 0 0 0.1rem var(--ld-col-rblck1);

    > input {
      color: var(--ld-col-rblck2);
      caret-color: transparent;

      &::placeholder {
        opacity: 0.5;
      }
    }
  }
  &:focus-within {
    box-shadow: inset 0 0 0 0.1rem var(--ld-col-rb-default);

    &:not([disabled]):not([aria-disabled]) {
      background-color: var(--ld-col-wht);
    }

    > input {
      outline: none;
    }

    .ld-theme-solvent &,
    .ld-theme-bubblegum &,
    .ld-theme-shake & {
      box-shadow: inset 0 0 0 0.1rem var(--ld-col-rp-default);
    }
    .ld-theme-tea & {
      box-shadow: inset 0 0 0 0.1rem var(--ld-col-rg-default);
    }
  }
}

.ld-input--invalid {
  box-shadow: inset 0 0 0 0.1rem var(--ld-col-rr-default);

  &:where(:not(:focus)) {
    color: var(--ld-col-rr-default);
    background-color: var(--ld-col-rr1);

    &::placeholder {
      color: var(--ld-col-rr3);
    }
  }
}
