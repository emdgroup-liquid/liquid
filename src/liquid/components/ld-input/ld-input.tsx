import { Component, Element, h, Host, Method, Prop } from '@stencil/core'
import { cloneAttributes } from '../../utils/cloneAttributes'

/**
 * The `ld-input` component. You can use it in conjunction with the `ld-label`
 * and the `ld-input-message` component. See examples in the docs for a better
 * understanding on how they can be used together.
 *
 * @slot start - The purpose of this slot is to add icons or buttons
 * to the input, __justifying the item to the end of the component__.
 * Styling for `ld-icon` and `ld-button` is provided within the `ld-input` component.
 * If you choose to place something different into the slot, you will probably
 * need to adjust some styles on the slotted item in order to make it fit right.
 * @slot end - The purpose of this slot is to add icons or buttons
 * to the input, __justifying the item to the start of the component__.
 * Styling for `ld-icon` and `ld-button` is provided within the `ld-input` component.
 * If you choose to place something different into the slot, you will probably
 * need to adjust some styles on the slotted item in order to make it fit right.
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part input - Actual input/textarea element
 * @part placeholder - Placeholder rendered for input type "file"
 */
@Component({
  tag: 'ld-input',
  styleUrl: 'ld-input.css',
  shadow: true,
})
export class LdInput implements InnerFocusable {
  @Element() el: HTMLInputElement | HTMLTextAreaElement
  private input: HTMLInputElement | HTMLTextAreaElement

  /** Input tone. Use `'dark'` on white backgrounds. Default is a light tone. */
  @Prop() tone: 'dark'

  /** The input value. */
  @Prop({ mutable: true, reflect: true }) value: string

  /** Set this property to `true` in order to mark the field visually as invalid. */
  @Prop() invalid: boolean

  /** The input placeholder. */
  @Prop() placeholder: string

  /** The input type. */
  @Prop() type: string

  /** Size of the input. */
  @Prop() size?: 'sm' | 'lg'

  /**
   * Uses textarea instead of input internally. Setting this attribute to true
   * disables the attribute type and both slots.
   */
  @Prop() multiline: boolean

  /**
   * Sets focus on the input
   */
  @Method()
  async focusInner() {
    if (this.input !== undefined) {
      this.input.focus()
    }
  }

  componentWillLoad() {
    this.el.querySelectorAll('ld-button').forEach((button) => {
      if (this.size !== undefined) {
        button.setAttribute('size', this.size)
      } else {
        button.removeAttribute('size')
      }
    })
    this.el.querySelectorAll('.ld-button').forEach((button) => {
      if (this.size === 'sm') {
        button.classList.remove('ld-button--lg')
        button.classList.add('ld-button--sm')
      } else if (this.size === 'lg') {
        button.classList.remove('ld-button--sm')
        button.classList.add('ld-button--lg')
      } else {
        button.classList.remove('ld-button--sm', 'ld-button--lg')
      }
    })
    this.el.querySelectorAll('ld-icon').forEach((icon) => {
      if (this.size !== undefined) {
        icon.setAttribute('size', this.size)
      } else {
        icon.removeAttribute('size')
      }
    })
    this.el.querySelectorAll('.ld-icon').forEach((icon) => {
      icon.classList.remove('ld-icon--sm', 'ld-icon--lg')
      if (this.size === 'sm') {
        icon.classList.remove('ld-icon--lg')
        icon.classList.add('ld-icon--sm')
      } else if (this.size === 'lg') {
        icon.classList.remove('ld-icon--sm')
        icon.classList.add('ld-icon--lg')
      } else {
        icon.classList.remove('ld-icon--sm', 'ld-icon--lg')
      }
    })
  }

  private handleBlur = (ev: FocusEvent) => {
    setTimeout(() => {
      this.el.dispatchEvent(ev)
    })
  }

  private handleFocus = (ev: FocusEvent) => {
    setTimeout(() => {
      this.el.dispatchEvent(ev)
    })
  }

  private handleInput() {
    if (!(this.input.getAttribute('aria-disabled') === 'true')) {
      this.value = this.input.value
      return
    }
    this.input.value = this.value || ''
  }

  private handleClick = (ev: MouseEvent) => {
    const target = ev.target as HTMLElement

    if (target.closest('ld-button')) return

    if (target === this.el) {
      this.input.focus()
      this.input.dispatchEvent(new Event('click', { bubbles: false }))
    } else {
      this.input.focus()
    }
  }

  render() {
    let cl = 'ld-input'
    if (this.size) cl += ` ld-input--${this.size}`
    if (this.tone) cl += ` ld-input--${this.tone}`
    if (this.invalid) cl += ' ld-input--invalid'

    if (this.multiline) {
      return (
        <Host class={cl} onClick={this.handleClick}>
          <textarea
            onBlur={this.handleBlur}
            onFocus={this.handleFocus}
            onInput={this.handleInput.bind(this)}
            part="input focusable"
            placeholder={this.placeholder}
            ref={(el) => (this.input = el)}
            {...cloneAttributes(this.el)}
            value={this.value}
          />
          {this.type === 'file' && (
            <span class="ld-input__placeholder" part="placeholder">
              {this.input?.value || this.placeholder}
            </span>
          )}
        </Host>
      )
    }

    return (
      <Host class={cl} onClick={this.handleClick}>
        <slot name="start"></slot>
        <input
          onBlur={this.handleBlur}
          onFocus={this.handleFocus}
          onInput={this.handleInput.bind(this)}
          part="input focusable"
          placeholder={this.placeholder}
          ref={(el) => (this.input = el)}
          type={this.type}
          {...cloneAttributes(this.el)}
          value={this.value}
        />
        {this.type === 'file' && (
          <span class="ld-input__placeholder" part="placeholder">
            {this.input?.value || this.placeholder}
          </span>
        )}
        <slot name="end"></slot>
      </Host>
    )
  }
}
