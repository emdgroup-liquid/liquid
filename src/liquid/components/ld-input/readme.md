---
eleventyNavigation:
  key: Input
  parent: Components
layout: layout.njk
title: Input
permalink: liquid/components/ld-input/
---

# ld-input

---

## Web component

{% example %}
<ld-input placeholder="Placeholder"></ld-input>
<ld-input value="Value"></ld-input>
{% endexample %}

### Disabled

{% example %}
<ld-input placeholder="Placeholder" disabled></ld-input>
<ld-input disabled value="Value"></ld-input>
{% endexample %}

### Light mode

{% example %}
<ld-input mode="light" placeholder="Placeholder"></ld-input>
<ld-input mode="light" value="Value"></ld-input>
{% endexample %}

### Light mode, disabled

{% example %}
<ld-input mode="light" disabled placeholder="Placeholder"></ld-input>
<ld-input mode="light" disabled value="Value"></ld-input>
{% endexample %}

### With label

{% example %}
<ld-label>
  Email Address
  <ld-input placeholder="jane.doe@example.com"></ld-input>
</ld-label>
{% endexample %}

Please reffer to the [ld-label](/liquid/components/ld-label/) docs for more information on the label component.

### With label and input message

{% example %}
<div style="display: grid; gap: 1rem; grid-template-columns: repeat(auto-fit, minmax(12rem, 1fr)); width: 100%;">
  <ld-label>
    Email Address
    <ld-input invalid placeholder="jane.doe@example.com" value="yolo"></ld-input>
    <ld-input-message>The email address is invalid.</ld-input-message>
  </ld-label>

  <ld-label>
    Password
    <ld-input type="password" value="asdf1234" type password></ld-input>
    <ld-input-message mode="info">Use at least one special character (~!@#$%^&*_-+=|\(){}[]:;<>,.?/)</ld-input-message>
  </ld-label>
</div>
{% endexample %}

#### On input message width and placement

By default, the input field stretches to the maximum width of its wrapping label, which in turn stretches to the width of its content (the computed width of the `ld-input-message` component). So, if you want to set certain constraints on the width of the all three components, you can do so by setting the constraint on the wrapping element – the `ld-label` component.

{% example %}
<ld-label>
  Email Address
  <ld-input placeholder="jane.doe@example.com"></ld-input>
  <ld-input-message mode="info">This info message is extremely long and makes all three components (the label, the input and itself) grow horizontaly.</ld-input-message>
</ld-label>

<ld-label style="max-width: 20rem">
  Email Address
  <ld-input placeholder="jane.doe@example.com"></ld-input>
  <ld-input-message mode="info">This info message is also extremely long, but since the label has a max width, all three components (the label, the input and itself) can take only the maximum width of the label.</ld-input-message>
</ld-label>
{% endexample %}

When displaying input messages conditionally (i.e. an error message becomes visible as soon as an input has been interacted with, but the value is still invalid) you should try to position UI elements in a way that prevents [layout shifts](https://web.dev/cls/). For instance, you can “reserve space” for your messages and then make them appear in the reserved space without pushing other content to the bottom (i.e. using `position: absolute` or some “flexy” layout). Needless to say, results look best if you keep the messages short.

### Example vanilla JS form validation

{% example %}
<style>
#example-form {
  display: flex;
  flex-wrap: wrap;
  gap: 1rem;
  width: 100%;
}
#example-form > * {
  align-self: flex-end;
  flex: 1 0 auto;
}
#example-form ld-button {
  margin-bottom: 1.6rem;
}
@media (max-width: 52rem) {
  #example-form > * {
    width: 100%;
  }
}
#example-form ld-input-message {
  visibility: hidden;
}
#example-form ld-input-message[covert="false"] {
  visibility: inherit;
}
</style>
<form id="example-form" novalidate>
  <ld-label>
    Login*
    <ld-input required placeholder="login"></ld-input>
    <ld-input-message visible="false">This field is required.</ld-input-message>
  </ld-label>
  <ld-label>
    Password*
    <ld-input required placeholder="password" type="password"></ld-input>
    <ld-input-message visible="false">This field is required.</ld-input-message>
  </ld-label>
  <ld-button>Submit</ld-button>
</form>
<script>
  const form = document.getElementById('example-form')
  const username = document.querySelector('#example-form ld-label:first-of-type ld-input')
  const usernameErrorMessage = document.querySelector('#example-form ld-label:first-of-type ld-input-message')
  const password = document.querySelector('#example-form ld-label:last-of-type ld-input')
  const passwordErrorMessage = document.querySelector('#example-form ld-label:last-of-type ld-input-message')
  async function validateInput(ldInput, ldInputMessage) {
    const value = await ldInput.getValue()
    if (!value) {
      ldInput.setAttribute('invalid', 'true')
      ldInputMessage.setAttribute('covert', 'false')
    } else {
      ldInput.removeAttribute('invalid')
      ldInputMessage.setAttribute('covert', 'true')
    }
  }
  username.addEventListener('input', ev => {
    validateInput(username, usernameErrorMessage)
  })
  username.addEventListener('blur', ev => {
    validateInput(username, usernameErrorMessage)
  })
  password.addEventListener('blur', ev => {
    validateInput(password, passwordErrorMessage)
  })
  password.addEventListener('input', ev => {
    validateInput(password, passwordErrorMessage)
  })
</script>
{% endexample %}


<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description                                                                        | Type                | Default     |
| --------- | --------- | ---------------------------------------------------------------------------------- | ------------------- | ----------- |
| `invalid` | `invalid` | Set this property to `true` in order to mark the field visually as invalid.        | `boolean`           | `undefined` |
| `mode`    | `mode`    | Input mode. Use `'dark'` on white backgrounds, use `'light'` on other backgrounds. | `"dark" \| "light"` | `'dark'`    |


## Methods

### `getValue() => Promise<string>`

Returns input value.

#### Returns

Type: `Promise<string>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
