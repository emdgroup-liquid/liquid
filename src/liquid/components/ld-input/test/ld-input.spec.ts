import { newSpecPage } from '@stencil/core/testing'
import { LdInput } from '../ld-input'

describe('ld-input', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input />`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders as dark input with prop tone set to "dark"', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input tone="dark" />`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders with value', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input value="yada-yada" />`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders type file', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input type="file" />`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders type file (multiline)', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input type="file" multiline />`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('updates value prop on value change', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input value="yada-yada" />`,
    })
    const ldInput = page.root
    expect(ldInput.value).toBe('yada-yada')

    const input = ldInput.shadowRoot.querySelector('input')
    expect(input.value).toBe('yada-yada')

    input.value = 'yoda-yoda'
    input.dispatchEvent(new Event('input'))
    await page.waitForChanges()

    expect(ldInput).toMatchSnapshot()
  })

  it('emits focus and blur event', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input />`,
    })
    const ldInput = page.root
    const input = page.root.shadowRoot.querySelector('input')

    const handlers = {
      onFocus() {
        return
      },
      onBlur() {
        return
      },
    }

    const spyFocus = jest.spyOn(handlers, 'onFocus')
    ldInput.addEventListener('focus', handlers.onFocus)
    input.dispatchEvent(new Event('focus'))
    jest.advanceTimersByTime(0)
    expect(spyFocus).toHaveBeenCalled()

    const spyBlur = jest.spyOn(handlers, 'onBlur')
    ldInput.addEventListener('blur', handlers.onBlur)
    input.dispatchEvent(new Event('blur'))
    jest.advanceTimersByTime(0)
    expect(spyBlur).toHaveBeenCalled()
  })

  it('renders with slot start', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input><span slot="start">hi</span></ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders with slot end', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input><span slot="end">hello</span></ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders with both slots', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input><span slot="start">hi</span><span slot="end">hello</span></ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('focuses the input on click of non-interactive elment inside the component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input><span slot="end"><span id="banana">üçå</span></span></ld-input>`,
    })
    const ldInput = page.root
    const banana = ldInput.querySelector('#banana') as HTMLElement
    const input = ldInput.shadowRoot.querySelector('input')

    input.focus = jest.fn()
    banana.dispatchEvent(new Event('click', { bubbles: true }))
    ldInput.dispatchEvent(new Event('click'))

    expect(input.focus).toHaveBeenCalledTimes(2)
  })

  it('forwards click to input (default)', async () => {
    const { root } = await newSpecPage({
      components: [LdInput],
      html: `<ld-input />`,
    })
    const input = root.shadowRoot.querySelector('input')

    input.focus = jest.fn()
    const dispatchSpy = jest.spyOn(input, 'dispatchEvent')
    root.click()

    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ bubbles: false })
    )
  })

  it('forwards click to input (multiline)', async () => {
    const { root } = await newSpecPage({
      components: [LdInput],
      html: `<ld-input multiline />`,
    })
    const textarea = root.shadowRoot.querySelector('textarea')

    textarea.focus = jest.fn()
    const dispatchSpy = jest.spyOn(textarea, 'dispatchEvent')
    root.click()

    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ bubbles: false })
    )
  })

  it('allows to set inner focus (default)', async () => {
    const { root } = await newSpecPage({
      components: [LdInput],
      html: `<ld-input />`,
    })
    const input = root.shadowRoot.querySelector('input')

    input.focus = jest.fn()
    await root.focusInner()

    expect(input.focus).toHaveBeenCalled()
  })

  it('allows to set inner focus (multiline)', async () => {
    const { root } = await newSpecPage({
      components: [LdInput],
      html: `<ld-input multiline />`,
    })
    const textarea = root.shadowRoot.querySelector('textarea')

    textarea.focus = jest.fn()
    await root.focusInner()

    expect(textarea.focus).toHaveBeenCalled()
  })

  it('removes size from ld-icon web component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input>
        <ld-icon name="placeholder" size="sm" slot="start"></ld-icon>
        <ld-icon name="placeholder" size="lg" slot="end"></ld-icon>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('removes size from ld-icon css component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input>
        <svg class="ld-icon ld-icon--sm" slot="start"></svg>
        <svg class="ld-icon ld-icon--lg" slot="end"></svg>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('sets size on ld-icon web component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input size="sm">
        <ld-icon name="placeholder" slot="start"></ld-icon>
        <ld-icon name="placeholder" size="lg" slot="end"></ld-icon>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('sets size on ld-icon css component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input size="sm">
        <svg class="ld-icon" slot="start"></svg>
        <svg class="ld-icon ld-icon--lg" slot="end"></svg>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('removes size from ld-button web component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input>
        <ld-button size="sm" slot="start">X</ld-button>
        <ld-button size="lg" slot="end">Y</ld-button>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('removes size from ld-button css component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input>
        <button class="ld-button ld-button--sm" slot="start">X</button>
        <button class="ld-button ld-button--lg" slot="end">Y</button>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('sets size on ld-button web component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input size="sm">
        <ld-button slot="start">X</ld-button>
        <ld-button size="lg" slot="end">Y</ld-button>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('sets size on ld-button css component', async () => {
    const page = await newSpecPage({
      components: [LdInput],
      html: `<ld-input size="sm">
        <button class="ld-button" slot="start">X</button>
        <button class="ld-button ld-button--lg" slot="end">Y</button>
      </ld-input>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
