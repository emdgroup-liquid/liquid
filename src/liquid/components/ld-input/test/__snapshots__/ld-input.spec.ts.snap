// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-input removes size from ld-button css component 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <button class="ld-button" slot="start">
    X
  </button>
  <button class="ld-button" slot="end">
    Y
  </button>
</ld-input>
`;

exports[`ld-input removes size from ld-button web component 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <ld-button slot="start">
    X
  </ld-button>
  <ld-button slot="end">
    Y
  </ld-button>
</ld-input>
`;

exports[`ld-input removes size from ld-icon css component 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <svg class="ld-icon" slot="start"></svg>
  <svg class="ld-icon" slot="end"></svg>
</ld-input>
`;

exports[`ld-input removes size from ld-icon web component 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <ld-icon name="placeholder" slot="start"></ld-icon>
  <ld-icon name="placeholder" slot="end"></ld-icon>
</ld-input>
`;

exports[`ld-input renders 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
</ld-input>
`;

exports[`ld-input renders as dark input with prop tone set to "dark" 1`] = `
<ld-input class="ld-input ld-input--dark" tone="dark">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
</ld-input>
`;

exports[`ld-input renders type file (multiline) 1`] = `
<ld-input class="ld-input" multiline="" type="file">
  <mock:shadow-root><textarea part="input focusable"></textarea>
    <span class="ld-input__placeholder" part="placeholder"></span>
  </mock:shadow-root>
</ld-input>
`;

exports[`ld-input renders type file 1`] = `
<ld-input class="ld-input" type="file">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable" type="file">
    <span class="ld-input__placeholder" part="placeholder"></span>
    <slot name="end"></slot>
  </mock:shadow-root>
</ld-input>
`;

exports[`ld-input renders with both slots 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <span slot="start">
    hi
  </span>
  <span slot="end">
    hello
  </span>
</ld-input>
`;

exports[`ld-input renders with slot end 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <span slot="end">
    hello
  </span>
</ld-input>
`;

exports[`ld-input renders with slot start 1`] = `
<ld-input class="ld-input">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <span slot="start">
    hi
  </span>
</ld-input>
`;

exports[`ld-input renders with value 1`] = `
<ld-input class="ld-input" value="yada-yada">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable" value="yada-yada">
    <slot name="end"></slot>
  </mock:shadow-root>
</ld-input>
`;

exports[`ld-input sets size on ld-button css component 1`] = `
<ld-input class="ld-input ld-input--sm" size="sm">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <button class="ld-button ld-button--sm" slot="start">
    X
  </button>
  <button class="ld-button ld-button--sm" slot="end">
    Y
  </button>
</ld-input>
`;

exports[`ld-input sets size on ld-button web component 1`] = `
<ld-input class="ld-input ld-input--sm" size="sm">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <ld-button size="sm" slot="start">
    X
  </ld-button>
  <ld-button size="sm" slot="end">
    Y
  </ld-button>
</ld-input>
`;

exports[`ld-input sets size on ld-icon css component 1`] = `
<ld-input class="ld-input ld-input--sm" size="sm">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <svg class="ld-icon ld-icon--sm" slot="start"></svg>
  <svg class="ld-icon ld-icon--sm" slot="end"></svg>
</ld-input>
`;

exports[`ld-input sets size on ld-icon web component 1`] = `
<ld-input class="ld-input ld-input--sm" size="sm">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable">
    <slot name="end"></slot>
  </mock:shadow-root>
  <ld-icon name="placeholder" size="sm" slot="start"></ld-icon>
  <ld-icon name="placeholder" size="sm" slot="end"></ld-icon>
</ld-input>
`;

exports[`ld-input updates value prop on value change 1`] = `
<ld-input class="ld-input" value="yoda-yoda">
  <mock:shadow-root>
    <slot name="start"></slot>
    <input part="input focusable" value="yoda-yoda">
    <slot name="end"></slot>
  </mock:shadow-root>
</ld-input>
`;
